File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int rows = grid.size();
        int col = grid[0].size();
        int cou = 0;
        for(int r = 0;r<rows;r++){
            for(int c= 0;c<col;c++){
                if(grid[r][c] == '1'){ 
                    dfs(grid,r,c);
                    cou++;

                }
            }
        }

        return cou;
        
    }

    void dfs(vector<vector<char>> &grid,int r,int c){
        if(r<0||r>grid.size()-1||c<0||c>grid[0].size()-1||grid[r][c]!='1') return;

        grid[r][c] = '0';

        dfs(grid,r+1,c);
        dfs(grid,r-1,c);
        dfs(grid,r,c-1);
        dfs(grid,r,c+1);
    }
};

File: 0200-number-of-islands/0200-number-of-islands.java
class Solution {
    private void dfs(int i,int j,char[][] grid){
        if(i<0 || j>=grid[0].length || j<0 || i>=grid.length ||grid[i][j]=='0') return ;
        grid[i][j] = '0';

        
                    //top
                    dfs(i-1,j,grid);
                    //bottom
                    dfs(i+1,j,grid);
                    //left
                    dfs(i,j-1,grid);
                    //right
                    dfs(i,j+1,grid);
        
    }
    public int numIslands(char[][] grid) {
        int n = grid.length;
        int m  = grid[0].length;
        
        int count = 0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                if(grid[i][j]=='1'){
                    dfs(i,j,grid);
                    count++;
                }
            }
        }

        return count;
        
    }
}